Initial version of Dockerfile for Boolector (#56)
Boolector/boolector: contrib/docker/DOCKER.md

# Using Boolector with Docker

Boolector has support for executing Boolector "input files" (e.g., `.smt2`)
files via the use of a Docker container. The base container used is Alpine
Linux.

> This guide presumes you are familiar with both Docker and Boolector, and are

Reviewer: I'm not a native speaker, but here is another suggestion (not sure if presume is the right choice here?)

```suggestion
This guide assumes the reader is familiar with both Docker and Boolector, and is
```

Author: Done!

able to install the necessary prerequisites to run Docker.

## Building the Docker image

To build the Docker image locally, you can perform the following:

```
git clone https://github.com/Boolector/boolector.git
cd boolector/contrib/docker
docker build -t btor/btor .
```

This will then build you a fresh Docker image, with Boolector built inside.

## Using Docker to run `.smt2` files

As Boolector is able to read SMT2LIB files via `stdin`, this means that it is
extremely easy to run Boolector against a `.smt2` file. For example:

```
#
# Presuming you're at the root of a Boolector clone

> #

Reviewer: Maybe remove the empty comment lines (to make it more compact)?

Author: Done!

cat src/tests/log/modelgensmt227.smt2 | docker run -i --rm btor/btor -m
```

Alternatively, to see Boolector's help:

```
docker run -i --rm btor/btor --help
```

*Note*: the `ENTRYPOINT` specified in the Boolector `Dockerfile` tells it to run
`boolector` without any arguments. You can specify additional arguments to
Boolector via your Docker `run` command.

### Sharing files to the container

The Boolector Docker image (by default) does not have access to your "host"
file-system, which is why `cat` was necessary to pipe in the contents of your
`.smt2` file into the image.

If you wish to allow the containerised version of Boolector to able to read
files from the host file-system, you can use Docker's `-v` flag to mount a
shared volume:

```
#
# Presuming you're at the root of a Boolector clone

> #

Reviewer: Ditto.

Author: Done!

docker run --rm -i -v $(pwd):/tmp btor/btor -m /tmp/src/tests/log/modelgensmt227.smt2
```

*Note*: the current working directory (`pwd`) has been mapped to `/tmp` within

> the container -- hence why the file-path to the SMT2LIB file you wish to run

Reviewer: ```suggestion
the container -- hence why the file-path to the SMT-LIB v2 file you wish to run
```

Author: Done!

starts with `/tmp`.

## Using Docker to work with Boolector's Python API

The Boolector Docker image has been built with support for Boolector's Python
API, however as the entry-point for the image is to run `boolector` (the
application), then slightly more work is required to work with Python inside of
the container:

```
#
# Presuming you're at the root of a Boolector clone

> #

Reviewer: Ditto.

Author: Done!

docker run -it -v $(pwd):/tmp --entrypoint=/bin/bash btor/btor -i

#
# You are now inside of the container!!!

> #

Reviewer: Ditto. Maybe merge with below comment?

Author: Done!

#
# Run a Python script using pyboolector

> #

Reviewer: Ditto.

Author: Done!

python /tmp/examples/api/python/api_usage_examples.py

#
# Leave the container!!!

> #

Reviewer: Ditto.

Author: Done!
